### PR TITLE
builtins: Implement ST_LineFromEncodedPolyline function

### LINKED ISSUE #48813: geo: implement ST_LineFromEncodedPolyline({text,int4}) -> geometry
Implement `ST_LineFromEncodedPolyline` on arguments {text,int4}, which should follow the [PostGIS implementation](https://postgis.net/docs/ST_LineFromEncodedPolyline.html).

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output manually. Tests here should just ensure the builtin is linked end to end.
* Ensure the documentation is regenerated by calling `make buildshort`.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48441 for an example PR.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-05-13T17:32:56Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by mknycha**
This function takes the encoded polyline and creates linestring out of it.
Works the same as function in PostGIS.
Release note (sql change): Implement geo builtin ST_LineFromEncodedPolyline.

Resolves #48813

### CODE REVIEW

File: pkg/geo/encode.go [hunk 1]
```diff
@@ -311,3 +312,57 @@ func StringToByteOrder(s string) binary.ByteOrder {
 		return DefaultEWKBEncodingFormat
 	}
 }
+
+// LineFromEncodedPolyline takes the encoded polyline ASCII and precision, decodes the points and returns them as a geometry
+func LineFromEncodedPolyline(encodedPolyline string, precision int) (Geometry, error) {
```
Comment (otan) [inline pkg/geo/encode.go:317]: Two things:
* Can you move the polyline logic (decodePolylinePoints, decodePointValue) to polyline.go.
* Can you rename this to ParseEncodedPolyline, and move this to parse.go?

File: pkg/geo/encode.go [hunk 2]
```diff
@@ -311,3 +312,57 @@ func StringToByteOrder(s string) binary.ByteOrder {
...
+	ls := geom.NewLineStringFlat(geom.XY, flatCoords).SetSRID(4326)
+
+	g, err := MakeGeometryFromGeomT(ls)
+	if err != nil {
+		return Geometry{}, fmt.Errorf("parsing geography error: %v", err)
+	}
+	return g, nil
+}
+
+func decodePolylinePoints(encoded string, precision int) []float64 {
```
Comment (otan) [inline pkg/geo/encode.go:328]: Can you add a comment to the function which describes how the algorithm works?

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 3]
```diff
@@ -4827,6 +4827,56 @@ The swap_ordinate_string parameter is a 2-character string naming the ordinates
...
+
+Returns valid results only if the polyline was encoded with 5 decimal places.
+
+See http://developers.google.com/maps/documentation/utilities/polylinealgorithm`,
+			}.String(),
+			Volatility: tree.VolatilityImmutable,
+		},
+		tree.Overload{
+			Types: tree.ArgTypes{
+				{"encoded polyline", types.String},
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:4856]: nit: encoded_polyline.

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 4]
```diff
@@ -4827,6 +4827,56 @@ The swap_ordinate_string parameter is a 2-character string naming the ordinates
 		},
 	),
 
+	"st_linefromencodedpolyline": makeBuiltin(defProps(),
+		tree.Overload{
+			Types: tree.ArgTypes{
+				{"encoded polyline", types.String},
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:4833]: nit: encoded_polyline.

File: pkg/geo/encode_test.go [hunk 5]
```diff
@@ -320,3 +320,101 @@ func TestSpatialObjectToGeoHash(t *testing.T) {
...
+			"_ibE?_ibE???",
+			1,
+			"SRID=4326;LINESTRING(0 10000, 0 20000, 0 20000)",
+		},
+		{
+			"EncodedPolyline of two points, integer values, precision 0",
+			"CACC",
+			0,
+			"SRID=4326;LINESTRING(1 2, 3 4)",
+		},
```
Comment (otan) [inline pkg/geo/encode_test.go:402]: for completelyness, can you add an empty string?

**Review by otan — COMMENTED** [review]
looking good! just some requested moves

File: pkg/geo/encode.go (hunk 1 above)
Comment (mknycha) [reply pkg/geo/encode.go:317]: Ok, done

File: pkg/geo/encode.go (hunk 2 above)
Comment (mknycha) [reply pkg/geo/encode.go:328]: Done

Comment (mknycha) [reply pkg/geo/encode.go:328]: But wait, the link to the algorithm is already added in the functions under `geo_builtins`. Do you want me to add it here anyway?

File: pkg/sql/sem/builtins/geo_builtins.go (hunk 3 above)
Comment (mknycha) [reply pkg/sql/sem/builtins/geo_builtins.go:4856]: Done

File: pkg/sql/sem/builtins/geo_builtins.go (hunk 4 above)
Comment (mknycha) [reply pkg/sql/sem/builtins/geo_builtins.go:4833]: Done

File: pkg/geo/encode_test.go (hunk 5 above)
Comment (mknycha) [reply pkg/geo/encode_test.go:402]: Good catch, added

**Review by otan — COMMENTED** [review]
i think this needs a rebase, otherwise good to go!@

**Comment by mknycha** [comment]
It looks like linter is telling me `Ensure generated code is up-to-date`, but it is up to date. I am also not able to replicate this problem locally with `make lint`. What can I do about it?

File: pkg/geo/polyline.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package geo
+
+import "math"
+
+// Reverse the polyline algorithm: https://developers.google.com/maps/documentation/utilities/polylinealgorithm
+func decodePolylinePoints(encoded string, precision int) []float64 {
```
[the remark this answers is not in the document]
Comment (mknycha) [reply pkg/geo/polyline.go:16]: Corrected

**Review by otan — APPROVED** [review]

**Pull request merged**